### ISSUE TITLE
Incorrect "incompatibility" when utilising `numbers` abstract base classes

### ISSUE THREAD

**Issue by yunruse** [microsoft/pyright#4616]
**Describe the bug**
Type-checking fails to capture the abstract base classes expressed in `numbers`.

**To Reproduce**
For example, the following code:

```python
from numbers import Rational

def my_exponentiation(x: Rational):
    pass

my_exponentiation(2)
```

has the following error reported:

```
Argument of type "Literal[2]" cannot be assigned to parameter "x" of type "Rational" in function "my_exponentiation"
  "Literal[2]" is incompatible with "Rational"   Pylance(reportGeneralTypeIssues)
```

**Expected behaviour**
`2` is an `int`, which is a `Rational`, so this should not occur.

By extension (see the example code below), the following subclass abstract base class relations should be understood (where here `<` indicates "is a subclass of"):
```
complex < numbers.Complex < numbers.Number
float < numbers.Real < numbers.Number
int < numbers.Integral < numbers.Rational < numbers.Number
fractions.Fraction < numbers.Rational
```

**Screenshots or Code**
The following code encapsulates all the errors:
```python
from numbers import Complex, Integral, Rational, Number, Real

def gives_number() -> float | int | complex:
    return 1

def takes_abc(x: Rational | Number | Complex | Real | Integral):
    pass

takes_abc(gives_number())
```
The same bug occurs in any subset of `gives_number` output or `takes_abc` input, giving such errors as `"complex" is incompatible with "Complex"` or `"float" is incompatible with "Number"`.

**VS Code extension or command-line**
I'm using Python 3.11, **VS Code with Pylance v2023.2.20**. This error occurs when type checking is set to basic.

**Additional context**

**The scope of this bug is _not_ a support of arbitrary abstract base classes!**
That would be a feature request, after all, and besides, my code only uses `numbers`.

Many thanks for your time! ^^

**Comment by erictraut** [comment]
Pyright is correct in emitting an error here. In the Python type system, the type `int` is not a subtype of `Rational`, so they are not type compatible. Not surpisingly, mypy generates the same error in this case.

There have been long discussions in the Python typing community (I can't locate them at the moment) about abstract numeric types like `Rational`, `Number`, etc. and rationalizing them with the concrete builtin numeric types, but there has been no resolution to this discussion. One potential resolution is to redefine `Rational`, etc. as Protocol classes (i.e. structural type definitions) rather than traditional nominal (named) classes.

In any case, pyright is working correctly here.
